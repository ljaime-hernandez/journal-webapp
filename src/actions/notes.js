import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";


export const startNewNote = () => {
    return async(dispatch, getState) => {

        // retrieves the user id from the store
        const uid = getState().auth.uid;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        // asynchronous function pushing a new note into the users journal
        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

        // we both assign the new note as active on our notesReducer and set it
        // ready on the notesScreen component for us to input the information
        dispatch(activeNote(doc.id, newNote));
        dispatch(addNewNote(doc.id, newNote));
    }
}

// sets the note into an active state, used in the NoteScreen component
export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id, 
        ...note
    }
})

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id, 
        ...note
    }
})

// uses the loadNotes helper to retrieve an array of notes from the 
// users collection for it to be rendered on the sidebar
export const startLoadingNotes = (uid) => {
    return async(dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));

    } 
}

// sends an array of notes to the notesReducer for them to be loaded
export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => {
    return async(dispatch, getState) => {

        // to save the note we retrieve the uid from the actual state on the
        // user store, as some users will forget or wont include a picture
        // on the note, then we make a small condition eliminating the
        // url value from the note
        const { uid } = getState().auth;

        if(!note.url){
            delete note.url;
        }

        // we declare the new note saving on a new variable, then proceeding for us to
        // delete the id, this is done so we reuse the previous note id to update it as
        // the note exists
        const noteToFirestore = {...note};
        delete noteToFirestore.id;

        // we make the request to the firebase db using the users uid, the note id and updating
        // all the items on its document with the update method
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

        // the refresh note will trigger an update on all the existing notes so they are
        // arranged accordingly and they also display the appropriate information on the screen,
        // afterwards sending us a screen popup with Swal to confirm the note was succesfully saved
        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved', note.title, 'success');
    } 
}

export const refreshNote = (id, note) => ({

    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const fileUploading = (file) => {
    return async (dispatch, getState) => {
        
        // the fileUploading action can be called only on an active note, either it is
        // a new note or an existing one, so we use the getState method to retrieve
        // the active note information from the store
        const {active:activeNote} = getState().notes;

        // a quick pop up will show using Swal with a loading spinner, while the
        // file is being uploaded using the fileUpload helper which will return the URL
        // where the file is uploaded, followed we call the startSaveNote action which
        // will save the whole note to our firebase collection and finally the
        // pop up will close after the process is done
        Swal.fire({
            title: 'Uploading... please wait',
            text: 'Please wait...',
            allowOutsideClick: false,
            willOpen: () => {
                Swal.showLoading();
            }
        });
        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote))

        Swal.close();
    }
}

export const startDelete = (id) => {
    return async(dispatch, getState) => {

        // similar to our previous actions, we use the getState method to retrieve the
        // user uid, then we use both the note id and the user uid to access to its
        // collection in firebase and trigger the delete method, deleting the document
        // stored on it
        const uid =  getState().auth.uid;

        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const noteLogout = () => {
    
    return {
        type: types.notesLogoutClean
    }
}