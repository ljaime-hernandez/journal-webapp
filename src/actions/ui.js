import { types } from "../types/types";

export const setError = (error) => {
    return {
        type: types.uiSetError,
        payload: error
    }
}

export const removeError = () => {
    return {
        type: types.uiRemoveError,
    }
}

export const startLoading = () => {
    return {
        type :types.uiStartLoading
    }
}

export const finishLoading = () => {
    return {
        type :types.uiFinishLoading
    }
}