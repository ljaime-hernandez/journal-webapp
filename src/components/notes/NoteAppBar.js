import React from 'react';

export const NoteAppBar = () => {
  return (
    <div className="notes__appbar">
        <span>Sept 19 1992</span>

        <div>
            <button className="btn">
                Picture    
            </button>
            
            <button className="btn">
                Save    
            </button>
        </div>
    </div>
  )
};
