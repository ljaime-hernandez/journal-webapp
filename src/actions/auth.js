import { firebase, googleAuthProvider } from '../firebase/firebaseConfig';
import { startLoading, finishLoading } from './ui';
import { types } from "../types/types";
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

// 
export const startLoginEmailPassword = (email, password) => {

    // the dispatch contains functions from different reducers, but the useDispatch hook
    // is linked to redux, using the environment developed in the store file along with
    // our redux functions, the reducers are contained in the combineReducer method, allowing
    // us to dispatch several actions at once
    return (dispatch) => {

        dispatch(startLoading());

        // account authentication from regular email accounts, it returns a callback which we will
        // use to dispatch the login action from our auth reducer and, either it goes thru or not
        // will also dispatch the finishLoading function to enable the LoginScreen button for login
        firebase.auth().signInWithEmailAndPassword(email, password)
        .then(({user}) => {
            dispatch(finishLoading());
            dispatch(login(user.uid, user.displayName));
        }).catch( e => {
            dispatch(finishLoading());
            Swal.fire('Error', e.message,'error');
        })
    }
}

export const startGoogleLogin = () => {

    // as this function is a callback, the dispatch will be sent back to our authReducer
    // environment for it to handle user login status
    return (dispatch) => {

        // the googleAuthProvider is an authentication method returned from our firebase
        // db, along with the signInWithPopup the user will be able to use its google
        // original account to create an account in our webpage, and for us to
        // use the information for authentication and also render a proper user interface
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
                )
            })
        }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const logout = () => {
    return {
        type: types.logout
    }
}

export const startLogout = () => {
    return (dispatch) => {
        firebase.auth().signOut();
        dispatch(noteLogout());
        dispatch(logout());
    }
}
    
export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
        .then( async ({user}) => {
            await user.updateProfile({displayName: name});

            dispatch(login(user.uid, user.displayName))
        }).catch( e => {
            console.log(e);
            // alert displaying an error message if theres errors on the user input
            // or if the email exists
            Swal.fire('Error', e.message,'error');
        })
    }
}

