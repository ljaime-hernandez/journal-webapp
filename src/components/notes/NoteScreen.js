import React from 'react';
import { NoteAppBar } from './NoteAppBar';

export const NoteScreen = () => {
  return (
    <div className="notes__main-content">

        <NoteAppBar/>

        <div className="notes__content">

        <input
            type="text"
            placeholder="Title of your journal"
            className="notes__title-input"
            autoComplete="off"
        />
        
        <textarea
            placeholder="What happened today?"
            className="notes__textarea"
        >   
        </textarea>

        <div className="notes__image">

        <img
            src="https://media.istockphoto.com/photos/panorama-of-the-foggy-winter-mountains-picture-id455011625?k=20&m=455011625&s=612x612&w=0&h=jkwMCTarHeDUvp3P1sD0dZ8hUrLMrDX3ICbBqP_E4FY="
            alt="landscape"
        />
        </div>

        </div>
    </div>
  )
};
