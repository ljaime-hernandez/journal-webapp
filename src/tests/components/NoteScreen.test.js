import thunk from "redux-thunk";
import configureStore from 'redux-mock-store';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { NoteScreen } from "../../components/notes/NoteScreen";
import { activeNote } from "../../actions/notes";

jest.mock('../../actions/notes', () => ({
    activeNote: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {
    auth: {
        uid:"HTXx027Mr1U94zHwXIsXNpT0M843",
        displayName:"Miguel"
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: {
            id: 'ABCDE',
            title: 'Test Title',
            body: 'Test Body',
            date: 0
        }
    }
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

describe('tests on NoteScreen component', () => {

    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <NoteScreen/>
            </MemoryRouter>
        </Provider>
        );

    test('should render properly', () => {
      
        expect(wrapper).toMatchSnapshot();
    });

    test('should launch the active note', () => {
      
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Title 2'
            }
        });

        expect(activeNote).toHaveBeenLastCalledWith(
            "ABCDE",
            {
                id: 'ABCDE',
                title: 'Title 2',
                body: 'Test Body',
                date: 0
            }
        );
    });
})