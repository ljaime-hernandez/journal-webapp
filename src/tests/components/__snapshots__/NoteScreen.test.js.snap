// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on NoteScreen component should render properly 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      Todays date: 
      February 11th 2022
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn btn-primary-notes"
        onClick={[Function]}
      >
        Add Picture
      </button>
      <button
        className="btn btn-primary-notes"
        onClick={[Function]}
      >
        Save Note
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Test Title"
      type="text"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="Test Body"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete Note
  </button>
</div>
`;
